### PR TITLE
native JS function compilation

### PR DESCRIPTION
**Pull request by littleredcomputer**
- move CSE to its own namespace to avoid the circular dependency compile->render->compile

- refactor JS rendering to allow compiler to use it

- adjust meaning of :native and :source compilation modes now you get what's compatible with your execution environment but you can also ask for a specific language, allowing tests to be bilingual

### CODE REVIEW

File: src/emmy/expression/compile.cljc
Comment (sritchie) [inline src/emmy/expression/compile.cljc:351]: I wonder if there is some way (not now, down the road) to use this feature in the TeX renderer? Is there some standard way to describe local bindings in TeX?

File: src/emmy/expression/compile.cljc
Comment (sritchie) [inline src/emmy/expression/compile.cljc:203]: I think this needs its leading backtick back

File: src/emmy/expression/compile.cljc
Comment (sritchie) [inline src/emmy/expression/compile.cljc:466]: good catch on my double into!

File: src/emmy/expression/compile.cljc
Comment (sritchie) [inline src/emmy/expression/compile.cljc:274]: did you mean to add a flatten? keyword arg?

File: src/emmy/expression/compile.cljc
Comment (sritchie) [inline src/emmy/expression/compile.cljc:308]: nice simplification, let's update this part of the docstring too.

File: src/emmy/expression/compile.cljc
Comment (sritchie) [inline src/emmy/expression/compile.cljc:328]: is there any benefit to using `let` here, or is it the other way around, that `var` has higher performance but potentially less safety?

File: src/emmy/expression/compile.cljc
Comment (sritchie) [inline src/emmy/expression/compile.cljc:344]: rather than compose with `js->clj`, what do you think of making an explicit call inside the ->JavaScript` function to a clojure vector constructor that we export? That way we could use this same function to compile functions that mutate the output vector, or return native js arrays by design. Thoughts?

File: src/emmy/expression/compile.cljc
Comment (sritchie) [inline src/emmy/expression/compile.cljc:353]: I'm quite happy that this seems to be unifying the different compile functions, between the compile-state-fn and compile-fn.

Do you think it makes sense to have a mode that compiles something that looks like a state derivative into a function of the parameters that returns a function of the local tuple, instead of flattening both levels into one function? Not for this PR, but I wanted to throw that into the design soup.

File: src/emmy/expression/render.cljc
Comment (sritchie) [inline src/emmy/expression/render.cljc:680]: Okay, I remember why I moved this function over. It seemed to me that this function was doing something sympy-like by taking an expression and trying to treat it as a function of the unspecified arguments.

In a similar way to the compile functions, it felt like it might be more consistent to have ->JavaScript take a function, and then generate the arguments and pass them into the fn instead of trying to infer and pull them out (which runs into argument order trouble, if you want to do something other than manually inspect the source before compiling.)

File: src/emmy/expression/render.cljc
Comment (sritchie) [inline src/emmy/expression/render.cljc:687]: I think this approach of with-out-str is fine for now, but I did hit some bugs a while ago where a println somewhere deep inside of emmy was triggering and getting into the string. No need to change it now, but I wanted to get that down somewhere!

**Review by sritchie — CHANGES_REQUESTED** [review]
This is looking great. Would you mind adding the notes from the PR to the CHANGELOG?

File: src/emmy/expression/compile.cljc
Comment (littleredcomputer) [reply src/emmy/expression/compile.cljc:351]: It could be done. In fact, it might even be cool. Something like (to borrow a Haskellism):
$$
\bf{let}
\qquad A_1 := x+y\cr
\qquad A_2 := \sin(x)\cr
\bf{in}
\qquad A_1^2 + A_2
$$
I don't know if that would actually render. I was sort of hoping GitHub would try to

File: src/emmy/expression/compile.cljc
Comment (littleredcomputer) [reply src/emmy/expression/compile.cljc]: done

File: src/emmy/expression/compile.cljc
Comment (littleredcomputer) [reply src/emmy/expression/compile.cljc]: no, this came out in the wash (fixed doc)

File: src/emmy/expression/compile.cljc
Comment (littleredcomputer) [reply src/emmy/expression/compile.cljc:327]: The [experts say](https://stackoverflow.com/questions/40070631/v8-javascript-performance-implications-of-const-let-and-var) it doesn't matter. I'm on the fence. Having been raised on const-correctness in ANSI C, I like to use const when possible, but if you write JS in modern style, practically everything is const except loop indices.

File: src/emmy/expression/compile.cljc
Comment (littleredcomputer) [reply src/emmy/expression/compile.cljc:343]: I put this in because I was getting a weird error of "IReduce.-reduce" not being present which I tracked down to the JS-nature of the returned array. This caused me to conclude that a JS array looked innocent but could fail to work in some distant context. Supplying a cli vector compatible constructor would do the trick, at the cost of the compiled code not looking like "real" JS. 

But you point out that this is a copy in performance critical code ... I'd rather "fix" this with an `aset` output convention than inject the constructor. I have personally benefited from the ability to paste the output into browser code, so I want to keep it.

File: src/emmy/expression/render.cljc
Comment (littleredcomputer) [reply src/emmy/expression/render.cljc:680]: We found the same thing. This is why I separated the functions via the map return value: nobody in compile-land looks at the inferred argument list, which is only used by the "printing" version, which I now agree is bogus (or at least that you have to manually deal with the function wrapping).

One way out of this would be to just have the ("legacy") `->JavaScript` just print the function body and leave the construction of the execution context up to the consumer (as compile.cljs does) instead of printing something that probably won't work

File: src/emmy/expression/render.cljc
Comment (littleredcomputer) [reply src/emmy/expression/render.cljc:687]: Good point. I heard a slight discordant note in the soundtrack as I used this feature and now repent of it. I'll fix it

**Comment by littleredcomputer** [comment]
CHANGELOG updated

**Comment by littleredcomputer** [comment]
ptal

**Comment by sritchie** [comment]
LGTM, once we get this merged I'll test it out with the animation code. So good!!

**Pull request merged**
